### ISSUE TITLE
Pure Objective-C test_spec fails to compile or run if it depends on Swift pod

### ISSUE THREAD

**Issue by adc-amatosov** [CocoaPods/CocoaPods#7170]
* [x] I've read and understood the [*CONTRIBUTING guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).

# Report
Pure Objective-C test_spec fails to compile or run if it depends on Swift pod (direct or transient dependency)

## What did you do?
Create lib with `pod lib create TestSpecs`
Configured test_spec for it using the Tests.m generated by `lib create`
Tried to compile and run tests
Complete project available at https://github.com/adc-amatosov/TestSpecs.git

## What did you expect to happen?
test_spec should generate compilable Test target even if it has only Objective-C files

## What happened instead?

In case of transient dependency build fails with error:
```
ld: library not found for -lswiftSwiftOnoneSupport
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

In case of direct dependency build succeeds but fails to run with error:
```
2017-10-26 12:25:50.737707-0700 xctest[85233:2589292] The bundle “TestSpecs-Unit-Tests.xctest” couldn’t be loaded because it is damaged or missing necessary resources. Try reinstalling the bundle.
2017-10-26 12:25:50.737829-0700 xctest[85233:2589292] (dlopen_preflight(~/Library/Developer/Xcode/DerivedData/TestSpecs-getnbftqhnqcyjaymnbrfnmdtskn/Build/Products/Debug-iphonesimulator/TestSpecs-Unit-Tests.xctest/TestSpecs-Unit-Tests): Library not loaded: @rpath/libswiftCore.dylib
  Referenced from: ~/Library/Developer/Xcode/DerivedData/TestSpecs-getnbftqhnqcyjaymnbrfnmdtskn/Build/Products/Debug-iphonesimulator/ReactiveCocoa/ReactiveCocoa.framework/ReactiveCocoa
  Reason: image not found)
Program ended with exit code: 82
```

## Workaround 
Adding an empty `dummy.swift` file to test_spec solves both issues

### Stack

```
   CocoaPods : 1.3.1
        Ruby : ruby 2.2.5p319 (2016-04-26 revision 54774) [x86_64-darwin15]
    RubyGems : 2.6.13
        Host : Mac OS X 10.13 (17A405)
       Xcode : 9.0 (9A235)
         Git : git version 2.10.2
Ruby lib dir : /Users/amatosov.ctr/.rbenv/versions/2.2.5/lib
Repositories : master - https://github.com/CocoaPods/Specs.git @ 7b6aee6a10e21aecd79278d52a33ed109809f197
```

### Installation Source

```
Executable Path: /Users/amatosov.ctr/.rbenv/versions/2.2.5/bin/pod
```

### Plugins

```
cocoapods-deintegrate : 1.0.1
cocoapods-packager    : 1.5.0
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.0.0
cocoapods-trunk       : 1.2.0
cocoapods-try         : 1.1.0
```

## Project that demonstrates the issue
https://github.com/adc-amatosov/TestSpecs.git
Branches:
 - `direct_swift_dep` for direct swift dependency case
 - `transient_swift_dep` for transient swift dependency
 - `workaround_with_dummy_swift` - a workaround that solves both cases

**Comment by dnkoutso** [comment]
Its definitely true, thanks for the report. Uncertain on the fix here. The code that does this is:
```
          def test_target_swift_debug_hack(test_target_bc)
            return unless test_target_bc.debug?
            return unless [target, *target.recursive_dependent_targets].any?(&:uses_swift?)
            ldflags = test_target_bc.build_settings['OTHER_LDFLAGS'] ||= '$(inherited)'
            ldflags << ' -lswiftSwiftOnoneSupport'
          end
```
Within `pod_target_installer.rb`.

It's because we are checking if any of the recursive dependent targets uses swift also.

**Comment by dnkoutso** [comment]
@segiddins should we instead when creating test targets:
```
              language = target.uses_swift? ? :swift : :objc
              native_test_target = project.new_target(product_type, name, platform_name, deployment_target, nil, language)
```
Also ask whether any of the recursive dependent targets `uses_swift?`?

**Comment by dnkoutso** [comment]
@adc-amatosov your workaround should be good for now.

**Comment by adc-amatosov** [comment]
If I remove `-lswiftSwiftOnoneSupport` build succeeds, but I start getting `Library not loaded: @rpath/libswiftCore.dylib` error.

**Comment by dnkoutso** [comment]
This might answer my question. @adc-amatosov can you directly edit your gem sources and edit file `pod_target_installer.rb` to use:

```ruby
language = [target, *target.recursive_dependent_targets].any?(&:uses_swift?) ? :swift : :objc
```
As a quick patch to see if that fixes the issue?

**Comment by dnkoutso** [comment]
accidental close.

**Comment by adc-amatosov** [comment]
nope. same  error
```
ld: library not found for -lswiftSwiftOnoneSupport
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by dnkoutso** [comment]
You are certain it used your patched version? One way to double check is make a typo and see if it crashes.

**Comment by adc-amatosov** [comment]
Yes. I verified this by doing a typo before saying it is not working.
Here is the change `pod_target_installer.rb` line 180
```
# language = target.uses_swift? ? :swift : :objc
language = [target, *target.recursive_dependent_targets].any?(&:uses_swift?) ? :swift : :objc
              
```

Here is how it changed generated Pods project:
https://github.com/adc-amatosov/TestSpecs/commit/6f6112f65e95f58b1b3a1dffcb0b04a3d5d6d2b9

**Comment by adc-amatosov** [comment]
Test targets are treated differently in Xcode and have a lot of holes in swift support.
From my past experience:
1. It will not link swift runtime unless you have at least 1 `.swift` file in the target. My workaround was to add `dummy.swift`
1. It will not link `libswift**.dylib` if you do not import them explicitly in your test target's `.swift` file (e.g. if the framework under test uses `CoreLocation`, running tests will fail with `Library not loaded: @rpath/libswiftCoreLocation.dylib`) the only workaround I have found was to put `import CoreLocation` in your `dummy.swift`

**Comment by dnkoutso** [comment]
@adc-amatosov so you mean Xcode in general without cocoapods requires a `.swift` file in order to add the swift libraries.

This sounds something like Xcode would do. In the meantime I still think the PR I have open is good, in fact I may need to improve it a bit.

I am not sure if ultimately we will be able to fix your issue directly from CocoaPods.

**Comment by dnkoutso** [comment]
@adc-amatosov I merged the other PR which is certainly better but overall still doesnt fix your problem 100%. I do not think we will be able to fix your problem 100%. so I will keep that as closed.

It seems the best viable workaround is to include a single `.swift` source to force xcode to link the swift sources.

If you feel strongly I can re-open but I do not see any quick fix here.

**Comment by anton-matosov** [comment]
> @adc-amatosov so you mean Xcode in general without CocoaPods requires a .swift file in order to add the swift libraries.

Correct. I had to do this with test targets created manually in Xcode. They were originally created as Objective-C and were using purely objective-c code for tests. But once I introduced 1 transient swift dependency they have start failing at runtime. So I had to add empty swift file to many of my test targets.

It will be really good if CocoaPods can give some handholding to users who are facing this kind of setup as it is pretty confusing and unclear how to solve.

Maybe give a hint/warning after `pod install` about this config or better install empty `dummy.swift` if CocoaPods can reliably detect setup of pure obj-c test target with swift in direct/transient dependencies and inform user that this workaround has been applied. 
Later on if Xcode team decide to fix the issue this workaround can be pulled out and after user re-runs `pod install` this `dummy.swift` will disappear.
IMO this is better than asking user to add it himself (as that's the only way to fix this) as it will be harder to clean it up later. I ended up adding 20+ dummy swift files just to fix this error.
